llama_hub/llama_packs/self_rag/base.py

```diff
@@ -0,0 +1,286 @@
+
+        Args:
+            paragraphs (List[str]): List of paragraphs to evaluate
+
+        Returns:
+            Tuple[Dict[int, float], Dict[int, str]]: Paragraphs final score and LLM predictions
+        """
+        paragraphs_final_score = {}
+        llm_response_text = {}
+
+        for p_idx, paragraph in enumerate(paragraphs):
+            pred = self.llm(paragraph, **self.generate_kwargs)
+            # Cache llm answer
+            llm_response_text[p_idx] = pred["choices"][0]["text"]
+
+            logprobs = pred["choices"][0]["logprobs"]
+            pred_log_probs = logprobs["top_logprobs"]
+            # Compute isRel score, on the first predicted token
+            isRel_score = _relevance_score(pred_log_probs[0])
+
+            # Compute isSup score
+            isSup_score = _is_supported_score(logprobs["tokens"], pred_log_probs)
+
+            # Compute isUse score
+            isUse_score = _is_useful_score(logprobs["tokens"], pred_log_probs)
+
+            paragraphs_final_score[p_idx] = (
+                isRel_score + isSup_score + 0.5 * isUse_score
+            )
+            if self.verbose:
+                print_text(
+                    f"Input: {paragraph}\nPrediction: {llm_response_text[p_idx]}\nScore: {paragraphs_final_score[p_idx]}\n",
+                    color="blue",
+                )
+                print_text(
+                    f"{p_idx + 1}/{len(paragraphs)} paragraphs done\n\n", color="blue"
+                )
+
+        return paragraphs_final_score, llm_response_text
+
+    def custom_query(self, query_str: str) -> Response:
+        """Run self-RAG."""
+        response = self.llm(prompt=_format_prompt(query_str), **_GENERATE_KWARGS)
+        answer = response["choices"][0]["text"]
+        if "[Retrieval]" in answer:
+            if self.verbose:
+                print_text("Retreival required\n", color="blue")
+            documents = self.retreiver.retrieve(query_str)
+            if self.verbose:
+                print_text(f"Received: {len(documents)} documents\n", color="blue")
+            paragraphs = [
+                _format_prompt(query_str, document.node.text) for document in documents
+            ]
+
+            if self.verbose:
+                print_text("Start evaluation\n", color="blue")
+            paragraphs_final_score, llm_response_text = self._run_critic(paragraphs)
+            if self.verbose:
+                print_text("End evaluation\n", color="blue")
+
+            best_paragraph_id = max(
+                paragraphs_final_score, key=paragraphs_final_score.get
+            )
+            answer = llm_response_text[best_paragraph_id]
+            if self.verbose:
+                print_text(f"Selected the best answer: {answer}\n", color="blue")
+
+        answer = _postprocess_answer(answer)
+        if self.verbose:
+            print_text(f"Final answer: {answer}\n", color="green")
+        return Response(response=str(answer))
```

Comment:
is there a way to add source_nodes too? given the nature of the self-rag algo, would be awesome to track the final retrieved nodes/paragraphs

Reply:
Added in the last commit